### PR TITLE
Long range transformers tip 

### PR DESCRIPTION
**Pull request by simondegheselle**
In this tip, we compare traditional transformers with novel efficient transformers. We'll use roBERTa as a baseline to compare against LongFormer and BigBird.

Notebook preview: https://colab.research.google.com/github/ml6team/quick-tips/blob/feature/long_seq_transformers/nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb

### CODE REVIEW

**Comment by simondegheselle** [comment]
@TDehaene ready for review 🕵️

**Comment by TDehaene** [comment]
> In this tip, we compare traditional transformers with novel efficient transformers. We'll use roBERTa as a baseline to compare against LongFormer and BigBird.
> 
> Notebook preview: https://colab.research.google.com/github/ml6team/quick-tips/blob/feature/long_seq_transformers/nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb

FWIW: I really like your PR opening statement: concise statement + a handy link to the colab wip notebook. kudos

File: nlp/2021_06_29_long_range_transformers/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Long range transformers
+
+In recent research there has been an extensive study for improving the calculation of attention in transformer architectures. Mostly for improving their capacity to handle longer token sequences. 👊
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/README.md:3]: Bit of a weird opening sentence, I sounds like there was this one multi-year study. I would rephrase it more to denote that it's a clear trend of the past year(s) or smting, in the opening statement

File: nlp/2021_06_29_long_range_transformers/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+The attention calculation is known to be quadratic in compatation time with respect to the sequence length 👎. These recent advances, however, are able to perform attention calculation in linear time with respect to the sequence length. This allows us to scale the transformer architecture such that it can handle input sequences beyond the usual token length of 512.
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/README.md:5]: typo: compatation > computation

Comment (TDehaene) [reply nlp/2021_06_29_long_range_transformers/README.md:5]: These recent advance > recent advances
we haven't shown work yet, so 'these' doesn't really refer to anything

Comment (TDehaene) [reply nlp/2021_06_29_long_range_transformers/README.md:5]: `ble to perform attention calculation in linear time`

weeeeel, not 100%, they aim to approach linear time

Comment (TDehaene) [reply nlp/2021_06_29_long_range_transformers/README.md:5]: `scale the transformer architecture such that it can handle` > `scale the transformer architecture so that it can handle`

Comment (TDehaene) [reply nlp/2021_06_29_long_range_transformers/README.md:5]: `that it can handle input sequences beyond the usual token length of 512` > `that it can more efficiently handle input sequences beyond the usual token length of 512`
The transformer doesn't inhibit lengths beyond 512, it's just not very usable

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 3]
```diff
@@ -0,0 +1,1940 @@
...
+    "Let's put these architectures to the test and see which one comes out on top 🏆!  \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "l8ymZTD_BSZS"
+   },
+   "source": [
+    "## 🛠️ Getting started: Install packages & download models\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:26]: well technically we're not downloading any models here

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 4]
```diff
@@ -0,0 +1,1940 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "UOdkK5pXBSZT"
+   },
+   "source": [
+    "## 💾 Dataset & downstream task\n",
+    "\n",
+    "We will use the [IMDB](https://huggingface.co/datasets/imdb) movie review dataset for binary sentiment classification. In the paper publication of LongFormer and BigBird, both architectures were compared against RoBERTa with this exact dataset.\n",
+    "*   13.6 % of the reviews are larger than 512 wordpieces \n",
+    "*   evaluation was done on the entire dataset and the performance improvement was .4% and .8% compared to the RoBERTa architecture for LongFormer and BigBird respectively 💪.\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:79]: - again, I would refrain from stating exact numbers in the text, as this might differ a tiny bit from run to run. 

- `evaluation was done on the entire dataset` > gives the impression that we evaluated on the training set > rephrase

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 5]
```diff
@@ -0,0 +1,1940 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {
+    "id": "t43_73glBSZU"
+   },
+   "outputs": [],
+   "source": [
+    "def get_deep_size(obj, seen=None):\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:279]: describe this method a bit (eg in a text cell above): what is this, why do we want to do this

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 6]
```diff
@@ -0,0 +1,1940 @@
...
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "1JYNVDC8BSZW",
+    "outputId": "0ad30f96-d5b9-4ec9-d3a9-8aa3ee3683b9"
+   },
+   "outputs": [],
+   "source": [
+    "train_ds, test_ds = filter_train_test_dataset_with_tokencount(ds['train'], ds['test'], token_count_cutoff)\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:261]: Don't really quite follow what we want to achieve here with the current description...
Why do we want equal n° of training samples?
Also: where do the values of 1270 and 1186 come from?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 7]
```diff
@@ -0,0 +1,1940 @@
...
+    "        per_device_eval_batch_size=4,\n",
+    "        evaluation_strategy=\"steps\",\n",
+    "        eval_steps=100, \n",
+    "        disable_tqdm=False,\n",
+    "        load_best_model_at_end=True,\n",
+    "        metric_for_best_model='accuracy', # default value is validation loss, we want the model with the highest accuracy \n",
+    "        warmup_steps=200,\n",
+    "        logging_steps=100,\n",
+    "        learning_rate=5e-6,\n",
+    "        fp16=True,\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:209]: did you notice a speedup with or without this?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 8]
```diff
@@ -0,0 +1,1940 @@
...
+    "### LongFormer"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "Njsdd-OaWBpW"
+   },
+   "outputs": [],
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:473]: has no output while the other two do, would be nice to harmonize this

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 9]
```diff
@@ -0,0 +1,1940 @@
...
+       "    <tr>\n",
+       "      <th>5</th>\n",
+       "      <td>roberta</td>\n",
+       "      <td>124.647170</td>\n",
+       "      <td>743692</td>\n",
+       "      <td>794.416961</td>\n",
+       "      <td>LONG</td>\n",
+       "      <td>0.485321</td>\n",
+       "      <td>0.899663</td>\n",
+       "      <td>0.900251</td>\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:814]: Given the evolution in F1 score, do you feel the model is sufficiently trained?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 10]
```diff
@@ -0,0 +1,1940 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {
+    "id": "eDI558ScBWpZ"
+   },
+   "outputs": [],
+   "source": [
+    "!pip install -q sklearn transformers datasets torch plotly"
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:42]: missing sentencepiece, one of steps fails since this wan't found (I think)

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 11]
```diff
@@ -0,0 +1,1940 @@
...
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "s63EKk7JVdsY",
+    "outputId": "5d2a0c41-0507-4f4d-b9be-1ae24019b74f"
+   },
+   "outputs": [],
+   "source": [
+    "ds = load_dataset('imdb')\n",
+    "\n",
+    "class IMDBDatasetTokenCountType(Enum):\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:105]: what is the purpose of this class? I could be wrong, but it feels a bit like unnecessary overhead

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 12]
```diff
@@ -0,0 +1,1940 @@
...
+    "        metric_for_best_model='accuracy', # default value is validation loss, we want the model with the highest accuracy \n",
+    "        warmup_steps=200,\n",
+    "        logging_steps=100,\n",
+    "        learning_rate=5e-6,\n",
+    "        fp16=True,\n",
+    "        dataloader_num_workers=0,\n",
+    "        output_dir='./{}-classification-consumer-complaints-results'.format(\n",
+    "            identifier),\n",
+    "        run_name='{}-classification-consumer-complaints'.format(identifier),\n",
+    "        logging_dir='./{}-classification-consumer-complaints-logging'.format(\n",
```
Comment (simondegheselle) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:214]: Should remove consumer-complaints suffix

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 13]
```diff
@@ -0,0 +1,1940 @@
...
+    "    \"\"\"Filter a Huggingface dataset on the length of the tokenized input text\"\"\"\n",
+    "    tokenizer =  RobertaTokenizer.from_pretrained('roberta-base')\n",
+    "    if tokencount_type == IMDBDatasetTokenCountType.LONG:\n",
+    "        return ds.filter(lambda x: len(tokenizer.batch_encode_plus([x['text']]).input_ids[0]) > 512)\n",
+    "    if tokencount_type == IMDBDatasetTokenCountType.SHORT:\n",
+    "        return ds.filter(lambda x: len(tokenizer.batch_encode_plus([x['text']]).input_ids[0]) < 512).filter(lambda x: len(tokenizer.batch_encode_plus([x['text']]).input_ids[0]) > 256)\n",
+    "    \n",
+    "def filter_train_test_dataset_with_tokencount(train_ds, test_ds, tokencount_type):\n",
+    "    filt_train_ds =  filter_dataset_with_tokencount(train_ds, tokencount_type)\n",
+    "    filt_test_ds = filter_dataset_with_tokencount(test_ds, tokencount_type)\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:119]: why are we filtering the testset

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 7 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:209]: Definitely, however, this doesn't work on Google Colab so maybe we should exclude it?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 6 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:261]: To have a fair comparison when comparing the impact of the token length
So that the model trained on short token length isn't performing better because there are more data samples to train on, do you think this is a valid approach? Maybe I could restructure it better to clarify the incentive?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 5 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:279]: Totally agree! Will include a description 👌

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 4 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:79]: Those are the statistical results mentioned in the paper, maybe I should clarify this with, "The BigBird & LongFormer papers state the following: ", or leave it out? what do you think?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 3 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:26]: Indeed 🙈

File: nlp/2021_06_29_long_range_transformers/README.md (hunk 2 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/README.md:5]: Thanks for the feedback Thomas! I'll change it

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 8 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:473]: +1

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 9 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:814]: Not sure tbf, I think it somehow is, because the validation loss is increasing over the final iterations. Should I retrain with for example 10 epochs to validate?

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 10 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:42]: Nice find! Thanks! I'll add it

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb (hunk 11 above)
Comment (simondegheselle) [reply nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:105]: We discussed this shortly in the call. Initially, it was to have a cleaner implementation for post-analysis (plotting, ...), but I agree this now seems a bit overkill since I am only using 2 enum types, I think a dictionary of

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 14]
```diff
@@ -0,0 +1,1940 @@
...
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "fig = go.Figure()\n",
+    "colors=[px.colors.qualitative.Plotly[1],px.colors.qualitative.Plotly[2],px.colors.qualitative.Plotly[8]]\n",
+    "\n",
+    "for index, model_name in enumerate(result['model_name'].unique()):\n",
+    "    trace_data = result[result.model_name == model_name].sort_values('max_token_length').reset_index()\n",
+    "    color_map = list(Color(colors[index]).range_to(Color('#000'), 75))\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:1859]: `Color` class unknown at this stage.
As a general tip, it's always a good idea to run your notebook end-to-end before submitting it, to doublecheck that nothing is missing

File: nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb [hunk 15]
```diff
@@ -0,0 +1,1940 @@
...
+    "    }\n",
+    "\n",
+    "def train_and_evaluate(model, identifier, tokenizer, train_ds, test_ds, max_length=1024):\n",
+    "    def tokenization(batched_text):\n",
+    "        return tokenizer(batched_text['text'], padding='max_length', truncation=True, max_length=max_length)\n",
+    "\n",
+    "    # tokenizing both the training and test dataset \n",
+    "    train_ds = train_ds.map(\n",
+    "        tokenization, batched=True, batch_size=len(train_ds), remove_columns=['text'])\n",
+    "    test_ds = test_ds.map(tokenization, batched=True,\n",
```
Comment (TDehaene) [inline nlp/2021_06_29_long_range_transformers/LongRangeTransformers.ipynb:187]: mapping the tokenization function cuts off the max_length of the testset as well. Do we want to do this for a certain reason?
